language/move-vm/runtime/src/loader.rs

```diff
@@ -711,14 +664,98 @@ impl Loader {
         })
     }
 
-    fn load_module(&self, id: &ModuleId, data_store: &mut impl DataStore) -> VMResult<Arc<Module>> {
+    // The process of loading is recursive, and module are cached by the loader as soon as
+    // they are verifiable (including dependencies).
+    // Effectively that means modules are cached from leaf to root in the dependency DAG.
+    // If a dependency error is found, loading stops and the error is returned.
+    // However all modules cached up to that point stay loaded.
+    // In principle that is not the safest model but it is justified
+    // by the fact that publishing is limited, and complex/tricky dependencies error
+    // are difficult or impossible to accomplish in reality.
+
+    // The process of deserialization and verification is not and it must not be under lock.
+    // So when publishing modules through the dependency DAG it may happen that a different
+    // thread had loaded the module after this process fetched it from storage.
+    // Caching will take care of that by asking for each dependency module again under lock.
+    fn load_module(
+        &self,
+        id: &ModuleId,
+        data_store: &mut impl DataStore,
+        possible_missing_modules: bool,
+    ) -> VMResult<Arc<Module>> {
+        // kept private to `load_module` to prevent verification errors from leaking
+        // and not being marked as invariant violations
+        fn deserialize_and_verify_module(
+            loader: &Loader,
+            bytes: Vec<u8>,
+            data_store: &mut impl DataStore,
+        ) -> VMResult<CompiledModule> {
+            let module = CompiledModule::deserialize(&bytes).map_err(|_| {
+                PartialVMError::new(StatusCode::CODE_DESERIALIZATION_ERROR)
+                    .finish(Location::Undefined)
+            })?;
+            loader.verify_module(&module)?;
+            let deps = module_dependencies(&module);
+            let loaded_deps =
+                loader.load_dependencies_expect_no_missing_modules(deps, data_store)?;
+            loader.verify_module_dependencies(&module, loaded_deps)?;
+            Ok(module)
+        }
+
         if let Some(module) = self.module_cache.lock().unwrap().module_at(id) {
             return Ok(module);
         }
-        let module = self.deserialize_and_verify_module(id, data_store)?;
+
+        let bytes = match data_store.load_module(id) {
+            Ok(bytes) => bytes,
+            Err(err) if possible_missing_modules => return Err(err),
+            Err(err) => {
+                crit!("[VM] Error fetching module with id {:?}", id);
+                return Err(err.expect_no_verification_errors());
+            }
+        };
+
+        let module = deserialize_and_verify_module(self, bytes, data_store)
+            .map_err(|e| e.expect_no_verification_errors())?;
         self.module_cache.lock().unwrap().insert(id.clone(), module)
     }
 
+    fn load_module_possible_missing_modules(
```

Comment:
I cannot stop laughing at those function names...

Reply:
oh I know lol
I couldn't think of anything succinct, so I tried as much as possible to make it *extra* clear what was going on. In doing so, it might actually be more confusing than something shorter.